File: solutions/javascript/difference-of-squares/1/difference-of-squares.js

// Using formulaes described at https://helloacm.com/the-difference-between-sum-of-squares-and-square-of-the-sum

export class Squares {
  constructor(n) {
    this._limit = n;
  }

  get sumOfSquares() {
    const n = this._limit;
    return (n * (n + 1) * ((n * 2) + 1)) / 6;
    

  }

  get squareOfSum() {
    const n = this._limit;
    return Math.pow((n * (n + 1)) / 2, 2);
  }

  get difference() {
    return this.squareOfSum - this.sumOfSquares;
  }
}

File: solutions/dart/difference-of-squares/1/lib/difference_of_squares.dart
class DifferenceOfSquares {
  int squareOfSum(int n) {
    int sum = n * (n + 1) ~/ 2;
    return sum * sum;
  }

  int sumOfSquares(int n) {
    return n * (n + 1) * (2 * n + 1) ~/ 6;
  }

  int differenceOfSquares(int n) {
    return squareOfSum(n) - sumOfSquares(n);
  }
}
